lib/spack/spack/test/conftest.py

```diff
@@ -121,34 +121,27 @@ def reset_compiler_cache():
     spack.compilers._compiler_cache = {}
 
 
-@pytest.fixture
-def clear_stage_root(monkeypatch):
-    """Ensure spack.stage._stage_root is not set at test start."""
-    monkeypatch.setattr(spack.stage, '_stage_root', None)
-    yield
-
-
 @pytest.fixture(scope='function', autouse=True)
-def mock_stage(clear_stage_root, tmpdir_factory, request):
+def mock_stage(tmpdir_factory, monkeypatch, request):
     """Establish the temporary build_stage for the mock archive."""
-    # Workaround to skip mock_stage for 'nomockstage' test cases
+    # The approach with this autouse fixture is to set the stage root
+    # instead of using spack.config.override() to avoid configuration
+    # conflicts with dozens of tests that rely on other configuration
+    # fixtures, such as config.
     if 'nomockstage' not in request.keywords:
+        # Set the build stage to the requested path
         new_stage = tmpdir_factory.mktemp('mock-stage')
         new_stage_path = str(new_stage)
 
-        # Set test_stage_path as the default directory to use for test stages.
-        current = spack.config.get('config:build_stage')
-        spack.config.set('config',
-                         {'build_stage': new_stage_path}, scope='user')
+        # Ensure the source directory exists within the new stage path
+        source_path = os.path.join(new_stage_path,
+                                   spack.stage._source_path_subdir)
+        mkdirp(source_path)
 
-        # Ensure the source directory exists
-        source_path = new_stage.join(spack.stage._source_path_subdir)
-        source_path.ensure(dir=True)
+        monkeypatch.setattr(spack.stage, '_stage_root', new_stage_path)
```

Comment:
(question) Is the `nomockstage` argument still required with this method of adjusting the stage path?

Reply:
The `mock_stage` fixture creates (during setup) and removes when present (during teardown) of the "well known" source path within the stage root. 

Given `mock_stage` is (now at) the function level and `autouse`, this file system activity is performed whether the test case needs it or not.  So I would think it would be better to either revisit having it `autouse` (which should be mirrored in the use of the `check_for_leftover_stage_files` fixture) OR make more use of the `nomockstage` mark to avoid the unnecessary overhead.

Reply:
BTW I meant "revisit" in a separate PR.

Reply:
> So I would think it would be better to either revisit having it autouse (which should be mirrored in the use of the check_for_leftover_stage_files fixture) OR make more use of the nomockstage mark to avoid the unnecessary overhead.

I prefer the former and agree that if it comes to this it ought to be dealt with in a later PR. However, I'm curious about:

> Given mock_stage is (now at) the function level and autouse, this file system activity is performed whether the test case needs it or not.

I was wondering if there might be a way to make the tests work regardless of whether this runs. The goal of this is to ensure that for the purposes of testing that the stages are created in a test-managed directory. I'm wondering if introducing an additional root path component would reconcile the concerns of the tests: e.g. right now the stage path read from `config.yaml` is relative to `/` but we could add a module-level variable to `Stage` that adjusts this root to be a temporary directory.

Does that seem reasonable? It could also be moved to another refactor PR but I prefer it to the approach of removing `autouse` or using `nomockstage`. I think it could be beneficial even in the case of removing the `autouse` but IMO it would also make it so `autouse` was only a performance problem.

Reply:
I wasn't the one to add `nomockstage` to address unit testing issues.  (Not sure I can reproduce their problem.)

I'll have to give this more thought.

Reply:
Created #13065

Reply:
OK that makes sense for this one, I pulled the PR and for the other tests I tried with/without `nomockstage` and it doesn't appear to be needed elsewhere.